llvm/lib/Transforms/Coroutines/CoroSplit.cpp

```diff
@@ -1455,6 +1462,64 @@ struct SwitchCoroutineSplitter {
     setCoroInfo(F, Shape, Clones);
   }
 
+  static Function *createNoAllocVariant(Function &F, coro::Shape &Shape,
+                                        SmallVectorImpl<Function *> &Clones) {
+    auto *OrigFnTy = F.getFunctionType();
+    auto OldParams = OrigFnTy->params();
+
+    SmallVector<Type *> NewParams;
+    NewParams.reserve(OldParams.size() + 1);
+    for (Type *T : OldParams) {
+      NewParams.push_back(T);
+    }
+    NewParams.push_back(PointerType::getUnqual(Shape.FrameTy));
+
+    auto *NewFnTy = FunctionType::get(OrigFnTy->getReturnType(), NewParams,
+                                      OrigFnTy->isVarArg());
+    Function *NoAllocF =
+        Function::Create(NewFnTy, F.getLinkage(), F.getName() + ".noalloc");
```

Comment:
I wonder if the function should be marked as `linkage = internal`. To my understanding, the `[[coro_must_elide]]` does not work across translation units, yet. As such, there won't be any cross-TU calls to the `.noalloc` function

Reply:
Does ThinLTO count as Cross-TU?

Reply:
Intuitively, I would think so, yes. But I never looked into ThinLTO before, so I don't know for sure.

Does #99285 enable `CoroAnnotationElide` already for ThinLTO?

Reply:
ThinLTO can import global values (variables and functions) from another modules. There would be a follow-up patch updating the pipeline so that coroutine passes are moved from pre-link to post-link. A previous attempt #90310 was blocked due to some conflict with Asan. I'll have a new version later.

Reply:
I think this should be marked as internal like .resume and .destroy function does. While LTO can import them, I don't feel we can benefit a lot from that.

Reply:
This pass is now configured to not run at all during LTO pre-link. So internal linkage SGTM.